File: c#/Canvas.Tests/ReplTest.cs
using System.Collections.Generic;
using Moq;
using NUnit.Framework;

namespace Canvas.Tests;

[TestFixture]
public class ReplTest
{
    private Mock<ICommandSource> _commandSource;
    private Mock<IDisplay> _display;

    [SetUp]
    public void Setup()
    {
        _commandSource = new Mock<ICommandSource>();
        _display = new Mock<IDisplay>();
    }

    [Test]
    public void CreateCanvasCommand()
    {
        FakeCommandSource(new CreateCanvas(1, 1));

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderPoints(new List<Point2D> {new(0, 0)});
    }

    [Test]
    public void InvalidCreateCanvasCommand()
    {
        FakeCommandSource(new CreateCanvas(-1, 1));

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderError();
    }

    [Test]
    public void PrintCanvasCommand()
    {
        FakeCommandSource(
            new CreateCanvas(1, 1),
            new PrintCanvas());

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderPoints(new List<Point2D> {new(0, 0)}, 2);
    }

    [Test]
    public void CreateCanvasCommandAgain()
    {
        FakeCommandSource(
            new CreateCanvas(1, 1),
            new CreateCanvas(0, 0),
            new PrintCanvas());

        new Repl(_commandSource.Object, _display.Object).Start();
        
        AssertExpectRenderPoints(new List<Point2D>(), 2);
    }

    [Test]
    public void OtherCommandFiredBeforeCreateCanvasCommand()
    {
        FakeCommandSource(new PrintCanvas());

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderError("create canvas first");
    }

    [Test]
    public void OtherCommandFiredAfterCanvasCreationFailed()
    {
        FakeCommandSource(
            new CreateCanvas(-1, 1),
            new PrintCanvas());

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderError();
    }

    [Test]
    public void DrawPointCommand()
    {
        FakeCommandSource(
            new CreateCanvas(1, 1),
            new DrawPoint(0, 0));

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderPoints(new List<Point2D> {new Border(0, 0)});
    }

    [Test]
    public void InvalidDrawPointCommand()
    {
        FakeCommandSource(new CreateCanvas(1, 1), new DrawPoint(-1, 0));

        new Repl(_commandSource.Object, _display.Object).Start();

        AssertExpectRenderError();
    }

    private void FakeCommandSource(params ICommand[] commands)
    {
        var moveNext = _commandSource.SetupSequence(c => c.MoveNext());
        var current = _commandSource.SetupSequence(c => c.Current);
        foreach (var command in commands)
        {
            moveNext.Returns(true);
            current.Returns(command);
        }
    }

    private void AssertExpectRenderPoints(List<Point2D> points, int times = 1)
    {
        _display.Verify(d => d.Render(points), Times.Exactly(times));
    }

    private void AssertExpectRenderError()
    {
        _display.Verify(d => d.RenderError(It.IsAny<string>()));
    }
    private void AssertExpectRenderError(string message)
    {
        _display.Verify(d => d.RenderError(message));
    }
}

File: f#/canvas/Canvas.Tests/ReplTest.fs
module Canvas.Tests.ReplTest

open Microsoft.FSharp.Core
open NUnit.Framework

type Command =
    | CreateCanvas of Width: int * Height: int
    | Quit

type Point = Point2D of X: int * Y: int
type CommandSource = unit -> Command
type Display =
    abstract Display: Set<Point> -> unit
    abstract Display: string -> unit

type Repl = CommandSource -> Display -> unit -> unit

let points =
    fun (width: int, height: int) ->
        set [ for x in 0 .. width - 1 do
                  for y in 0 .. height - 1 do
                      Point2D(x, y) ]

let repl: Repl =
    fun commandSource display ->
        match commandSource () with
        | CreateCanvas (w, h) -> fun () -> display.Display (points (w, h))
        | Quit -> fun() -> display.Display("Good bye!")

type MockDisplay () =
   interface Display with
      member this.Display(points: Set<Point>) = Assert.AreEqual(List<Point>.Empty, points)
      member this.Display(message: string) = Assert.AreEqual("Good bye!", message)

[<SetUp>]
let Setup () = ()

[<Test>]
let CreateCanvasCommand () =
    let commandSource = fun () -> CreateCanvas(Width = 0, Height = 0)    
    let start = repl commandSource (MockDisplay())
    start ()

[<Test>]
let QuitCommand () =
    let commandSource =fun () -> Quit    
    let start = repl commandSource (MockDisplay())
    start()